<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Generate a .Rmd file containing code to perform differential expression analysis with DESeq2 with custom model matrix — DESeq2.length.createRmd • compcodeR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Generate a .Rmd file containing code to perform differential expression analysis with DESeq2 with custom model matrix — DESeq2.length.createRmd"><meta property="og:description" content="A function to generate code that can be run to perform differential expression analysis of RNAseq data (comparing two conditions) using the DESeq2 package. The code is written to a .Rmd file. This function is generally not called by the user, the main interface for performing differential expression analysis is the runDiffExp function."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">compcodeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/compcodeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/phylocompcodeR.html">Including inter-species measurements in differential expression analysis of RNAseq data with the compcodeR package</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/csoneson/compcodeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Generate a <code>.Rmd</code> file containing code to perform differential expression analysis with DESeq2 with custom model matrix</h1>
    <small class="dont-index">Source: <a href="https://github.com/csoneson/compcodeR/blob/HEAD/R/generateRmdCodeDiffExpPhylo.R" class="external-link"><code>R/generateRmdCodeDiffExpPhylo.R</code></a></small>
    <div class="hidden name"><code>DESeq2.length.createRmd.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>A function to generate code that can be run to perform differential expression analysis of RNAseq data (comparing two conditions) using the DESeq2 package. The code is written to a <code>.Rmd</code> file. This function is generally not called by the user, the main interface for performing differential expression analysis is the <code><a href="runDiffExp.html">runDiffExp</a></code> function.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">DESeq2.length.createRmd</span><span class="op">(</span>
  <span class="va">data.path</span>,
  <span class="va">result.path</span>,
  <span class="va">codefile</span>,
  <span class="va">fit.type</span>,
  <span class="va">test</span>,
  beta.prior <span class="op">=</span> <span class="cn">TRUE</span>,
  independent.filtering <span class="op">=</span> <span class="cn">TRUE</span>,
  cooks.cutoff <span class="op">=</span> <span class="cn">TRUE</span>,
  impute.outliers <span class="op">=</span> <span class="cn">TRUE</span>,
  extra.design.covariates <span class="op">=</span> <span class="cn">NULL</span>,
  nas.as.ones <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>data.path</dt>
<dd><p>The path to a .rds file containing the <code>phyloCompData</code> object that will be used for the differential expression analysis.</p></dd>
<dt>result.path</dt>
<dd><p>The path to the file where the result object will be saved.</p></dd>
<dt>codefile</dt>
<dd><p>The path to the file where the code will be written.</p></dd>
<dt>fit.type</dt>
<dd><p>The fitting method used to get the dispersion-mean relationship. Possible values are <code>"parametric"</code>, <code>"local"</code> and <code>"mean"</code>.</p></dd>
<dt>test</dt>
<dd><p>The test to use. Possible values are <code>"Wald"</code> and <code>"LRT"</code>.</p></dd>
<dt>beta.prior</dt>
<dd><p>Whether or not to put a zero-mean normal prior on the non-intercept coefficients. Default is <code>TRUE</code>.</p></dd>
<dt>independent.filtering</dt>
<dd><p>Whether or not to perform independent filtering of the data. With independent filtering=TRUE, the adjusted p-values for genes not passing the filter threshold are set to NA.</p></dd>
<dt>cooks.cutoff</dt>
<dd><p>The cutoff value for the Cook's distance to consider a value to be an outlier. Set to Inf or FALSE to disable outlier detection. For genes with detected outliers, the p-value and adjusted p-value will be set to NA.</p></dd>
<dt>impute.outliers</dt>
<dd><p>Whether or not the outliers should be replaced by a trimmed mean and the analysis rerun.</p></dd>
<dt>extra.design.covariates</dt>
<dd><p>A vector containing the names of extra control variables to be passed to the design matrix of <code>DESeq2</code>. All the covariates need to be a column of the <code>sample.annotations</code> data frame from the <code><a href="phyloCompData.html">phyloCompData</a></code> object, with a matching column name. The covariates can be a numeric vector, or a factor. Note that "condition" factor column is always included, and should not be added here. See Details.</p></dd>
<dt>nas.as.ones</dt>
<dd><p>Whether or not adjusted p values that are returned as <code>NA</code> by <code>DESeq2</code> should be set to <code>1</code>. This option is useful for comparisons with other methods. For more details, see section "I want to benchmark DESeq2 comparing to other DE tools" from the <code>DESeq2</code> vignette (available by running <code><a href="https://cran.rstudio.com/web/packages/DESeq2/vignettes/DESeq2.html" class="external-link">vignette("DESeq2", package = "DESeq2")</a></code>). Default to <code>FALSE</code>.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>The function generates a <code>.Rmd</code> file containing the code for performing the differential expression analysis. This file can be executed using e.g. the <code>knitr</code> package.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For more information about the methods and the interpretation of the parameters, see the <code>DESeq2</code> package and the corresponding publications.</p>

<p>The lengths matrix is used as a normalization factor and applied to the <code>DESeq2</code>
model in the way explained in <code><a href="https://rdrr.io/pkg/DESeq2/man/normalizationFactors.html" class="external-link">normalizationFactors</a></code>
(see examples of this function).
The provided matrix will be multiplied by the default normalization factor 
obtained through the <code>estimateSizeFactors</code> function.</p>
<p>The <code>design</code> model used in the <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></code>
uses the "condition" column of the <code>sample.annotations</code> data frame from the <code><a href="phyloCompData.html">phyloCompData</a></code> object
as well as all the covariates named in <code>extra.design.covariates</code>.
For example, if <code>extra.design.covariates = c("var1", "var2")</code>, then
<code>sample.annotations</code> must have two columns named "var1" and "var2", and the design formula
in the <code><a href="https://rdrr.io/pkg/DESeq2/man/DESeqDataSet.html" class="external-link">DESeqDataSetFromMatrix</a></code> function will be:
<code>~ condition + var1 + var2</code>.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Anders S and Huber W (2010): Differential expression analysis for sequence count data. Genome Biology 11:R106</p>
<p>Love, M.I., Huber, W., Anders, S. (2014) Moderated estimation of fold change and dispersion for RNA-seq data with DESeq2. Genome Biology, 15:550. 10.1186/s13059-014-0550-8.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Charlotte Soneson, Paul Bastide, Mélina Gallopin</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span></span>
<span class="r-in"><span class="kw">if</span> <span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/mikelove/DESeq2" class="external-link">DESeq2</a></span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span class="r-in"><span class="va">tmpdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/normalizePath.html" class="external-link">normalizePath</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempdir</a></span><span class="op">(</span><span class="op">)</span>, winslash <span class="op">=</span> <span class="st">"/"</span><span class="op">)</span></span>
<span class="r-in"><span class="co">## Simulate data</span></span>
<span class="r-in"><span class="va">mydata.obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="generateSyntheticData.html">generateSyntheticData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"mydata"</span>, n.vars <span class="op">=</span> <span class="fl">1000</span>, </span>
<span class="r-in">                                    samples.per.cond <span class="op">=</span> <span class="fl">5</span>, n.diffexp <span class="op">=</span> <span class="fl">100</span>, </span>
<span class="r-in">                                    id.species <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,</span>
<span class="r-in">                                    lengths.relmeans <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1000</span><span class="op">)</span>,</span>
<span class="r-in">                                    lengths.dispersions <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/GammaDist.html" class="external-link">rgamma</a></span><span class="op">(</span><span class="fl">1000</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span class="r-in">                                    output.file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"mydata.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="co">## Add covariates</span></span>
<span class="r-in"><span class="co">## Model fitted is count.matrix ~ condition + test_factor + test_reg</span></span>
<span class="r-in"><span class="fu">sample.annotations</span><span class="op">(</span><span class="va">mydata.obj</span><span class="op">)</span><span class="op">$</span><span class="va">test_factor</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, each <span class="op">=</span> <span class="fl">5</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu">sample.annotations</span><span class="op">(</span><span class="va">mydata.obj</span><span class="op">)</span><span class="op">$</span><span class="va">test_reg</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html" class="external-link">rnorm</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">mydata.obj</span>, <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"mydata.rds"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="co">## Diff Exp</span></span>
<span class="r-in"><span class="fu"><a href="runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">tmpdir</span>, <span class="st">"mydata.rds"</span><span class="op">)</span>, result.extent <span class="op">=</span> <span class="st">"DESeq2"</span>, </span>
<span class="r-in">           Rmdfunction <span class="op">=</span> <span class="st">"DESeq2.length.createRmd"</span>, </span>
<span class="r-in">           output.directory <span class="op">=</span> <span class="va">tmpdir</span>, fit.type <span class="op">=</span> <span class="st">"parametric"</span>,</span>
<span class="r-in">           test <span class="op">=</span> <span class="st">"Wald"</span>,</span>
<span class="r-in">           extra.design.covariates <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/Vector-class.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"test_factor"</span>, <span class="st">"test_reg"</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="op">}</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Vector 'id.species' must be a factor. Transforming.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Error in sample.annotations(mydata.obj)$test_factor &lt;- factor(rep(1:2,  : </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   could not find function "sample.annotations"</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Charlotte Soneson, Paul Bastide, Mélina Gallopin.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.2.9000.</p>
</div>

      </footer></div>

  


  

  </body></html>

