<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Comparing methods for differential expression analysis of RNAseq data with the compcodeR package • compcodeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Comparing methods for differential expression analysis of RNAseq data with the compcodeR package">
<meta property="og:description" content="compcodeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">compcodeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.37.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/compcodeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/phylocompcodeR.html">Including inter-species measurements in differential expression analysis of RNAseq data with the compcodeR package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/csoneson/compcodeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Comparing methods for differential expression
analysis of RNAseq data with the compcodeR package</h1>
                        <h4 data-toc-skip class="author">Charlotte
Soneson</h4>
            
            <h4 data-toc-skip class="date">2023-09-28</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csoneson/compcodeR/blob/HEAD/vignettes/compcodeR.Rmd" class="external-link"><code>vignettes/compcodeR.Rmd</code></a></small>
      <div class="hidden name"><code>compcodeR.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/csoneson/compcodeR" class="external-link">compcodeR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: sm</span></span>
<span><span class="co">#&gt; Package 'sm', version 2.2-5.7: type help(sm) for summary information</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p><em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
is an R package that provides an interface to several popular methods
for differential expression analysis of RNAseq data and contains
functionality for comparing the outcomes and performances of several
differential expression methods applied to the same data set. The
package also contains a function for generating synthetic RNAseq counts,
using the simulation framework described in more detail in <span class="citation">Soneson and Delorenzi (2013)</span>.</p>
<p>This vignette provides a tutorial on how to use the different
functionalities of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package.</p>
<p>Currently, the differential expression interfaces provided in the
package are restricted to comparisons between two conditions. However,
many of the comparison functions are more general and can also be
applied to test results from other contrast types, as well as to test
results from other data types than RNAseq.</p>
<p><strong>Important!</strong> Since <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
provides interfaces to differential expression analysis methods
implemented in other R packages, take care to cite the appropriate
references if you use any of the interface functions (see the reference
manual for more information). Also be sure to always check the code that
was used to run the differential expression analysis, using e.g. the
<code><a href="../reference/generateCodeHTMLs.html">generateCodeHTMLs()</a></code> function (see <a href="#performing-differential-expression-analysis">below</a> for more
information) to make sure that parameters etc. agree with your
intentions and that there were no errors or serious warnings. These
interface functions are provided for convenience and as templates for
other, user-defined workflows, and there is no guarantee that the
included differential expression code is kept up-to-date with the latest
recommendations and best practices for running each of the interfaced
methods, or that the chosen settings are suitable in all situations, and
the user should make sure that the analysis is performed in the way they
intend. For instructions on how to add your own DE method, see <a href="#providing-your-own-differential-expression-code">below</a>.</p>
</div>
<div class="section level2">
<h2 id="the-compdata-class">The <code>compData</code> class<a class="anchor" aria-label="anchor" href="#the-compdata-class"></a>
</h2>
<p>Within the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package (version &gt;= 0.2.0), data sets and results are represented as
objects of the <code>compData</code> class. The functions in the package
are still compatible with the list-based representation used in version
0.1.0, but we strongly encourage users to use the <code>compData</code>
class, and all results generated by the package will be given in this
format. If you have a data or result object generated with <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
version 0.1.0, you can convert it to a <code>compData</code> object
using the <code>convertListTocompData</code> function.</p>
<p>A <code>compData</code> object has at least three slots, containing
the count matrix, sample annotations and a list containing at least an
identifying name and a unique ID for the data set. It can also contain
variable annotations, such as information regarding genes that are known
to be differentially expressed. After performing a differential
expression analysis, the <code>compData</code> object contains
additional information, such as which method was used to perform the
analysis, which settings were used and the gene-wise results from the
analysis. More detailed information about the <code>compData</code>
class are available in the sections on <a href="#the-data-object">the
data object</a> and <a href="#the-result-object">the result
object</a>.</p>
</div>
<div class="section level2">
<h2 id="a-sample-workflow">A sample workflow<a class="anchor" aria-label="anchor" href="#a-sample-workflow"></a>
</h2>
<p>This section contains a sample workflow showing the main
functionalities of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package. We start by <a href="#simulating-data">generating a synthetic
count data set</a>, to which we then <a href="#performing-differential-expression-analysis">apply three
different differential expression methods</a>. Finally, we <a href="#comparing-results-from-several-differential-expression-methods">compare
the outcome</a> of the three methods and generate a report summarizing
the results.</p>
<div class="section level3">
<h3 id="simulating-data">Simulating data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h3>
<p>The simulations are performed following the description by <span class="citation">Soneson and Delorenzi (2013)</span>. As an example, we
use the <code>generateSyntheticData</code> function to generate a
synthetic count data set containing 12,500 genes and two groups of 5
samples each, where 10% of the genes are simulated to be differentially
expressed between the two groups (equally distributed between up- and
downregulated in group 2 compared to group 1). Furthermore, the counts
for all genes are simulated from a Negative Binomial distribution with
the same dispersion in the two sample groups, and no outlier counts are
introduced. We filter the data set by excluding only the genes with zero
counts in all samples (i.e., those for which the total count is 0). This
simulation setting corresponds to the one denoted <span class="math inline">\(B_{625}^{625}\)</span> in <span class="citation">Soneson and Delorenzi (2013)</span>. The following code
creates a <code>compData</code> object containing the simulated data set
and saves it to a file named
<code>"B_625_625_5spc_repl1.rds"</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">B_625_625</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateSyntheticData.html">generateSyntheticData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"B_625_625"</span>, n.vars <span class="op">=</span> <span class="fl">12500</span>, </span>
<span>                                   samples.per.cond <span class="op">=</span> <span class="fl">5</span>, n.diffexp <span class="op">=</span> <span class="fl">1250</span>, </span>
<span>                                   repl.id <span class="op">=</span> <span class="fl">1</span>, seqdepth <span class="op">=</span> <span class="fl">1e7</span>, </span>
<span>                                   fraction.upregulated <span class="op">=</span> <span class="fl">0.5</span>, </span>
<span>                                   between.group.diffdisp <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>                                   filter.threshold.total <span class="op">=</span> <span class="fl">1</span>, </span>
<span>                                   filter.threshold.mediancpm <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                   fraction.non.overdispersed <span class="op">=</span> <span class="fl">0</span>, </span>
<span>                                   output.file <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1.rds"</span><span class="op">)</span></span></code></pre></div>
<p>The <code>summarizeSyntheticDataSet</code> function provides
functionality to check some aspects of the simulated data by generating
a report summarizing the parameters that were used for the simulation,
as well as including some diagnostic plots. The report contains two
MA-plots, showing the estimated average expression level and the
log-fold change for all genes, indicating either the truly
differentially expressed genes or the total number of outliers
introduced for each gene. It also shows the log-fold changes estimated
from the simulated data versus those underlying the simulation. The
input to the <code>summarizeSyntheticDataSet</code> function can be
either a <code>compData</code> object or the path to a file containing
such an object.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarizeSyntheticDataSet.html">summarizeSyntheticDataSet</a></span><span class="op">(</span>data.set <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1.rds"</span>, </span>
<span>                          output.filename <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1_datacheck.html"</span><span class="op">)</span></span></code></pre></div>
<p>Figure 1 shows two of the figures generated by this function. The top
panel shows an MA plot with the genes colored by the true differential
expression status. The bottom panel shows the relationship between the
true log-fold changes between the two sample groups underlying the
simulation, and the estimated log-fold changes based on the simulated
counts.</p>
<div class="figure" style="text-align: center">
<img src="compcodeR_check_figure/maplot-trueDEstatus.png" alt="Example figures from the summarization report generated for a simulated data set. The top panel shows an MA plot, with the genes colored by the true differential expression status. The bottom panel shows the relationship between the true log-fold changes between the two sample groups underlying the simulation, and the estimated log-fold changes based on the simulated counts. Also here, the genes are colored by the true differential expression status." width="324"><img src="compcodeR_check_figure/logfoldchanges.png" alt="Example figures from the summarization report generated for a simulated data set. The top panel shows an MA plot, with the genes colored by the true differential expression status. The bottom panel shows the relationship between the true log-fold changes between the two sample groups underlying the simulation, and the estimated log-fold changes based on the simulated counts. Also here, the genes are colored by the true differential expression status." width="324"><p class="caption">
Example figures from the summarization report generated for a simulated
data set. The top panel shows an MA plot, with the genes colored by the
true differential expression status. The bottom panel shows the
relationship between the true log-fold changes between the two sample
groups underlying the simulation, and the estimated log-fold changes
based on the simulated counts. Also here, the genes are colored by the
true differential expression status.
</p>
</div>
</div>
<div class="section level3">
<h3 id="performing-differential-expression-analysis">Performing differential expression analysis<a class="anchor" aria-label="anchor" href="#performing-differential-expression-analysis"></a>
</h3>
<p>We will now apply some of the interfaced differential expression
methods to find genes that are differentially expressed between the two
conditions in the simulated data set. This is done through the
<code>runDiffExp</code> function, which is the main interface for
performing differential expression analyses in <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>.
The code below applies three differential expression methods to the data
set generated above: the <code>voom</code> transformation from the
<em><a href="https://bioconductor.org/packages/3.18/limma" class="external-link">limma</a></em>
package (combined with <code>limma</code> for differential expression),
the exact test from the <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>
package, and a regular t-test applied directly on the count level
data.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1.rds"</span>, </span>
<span>           result.extent <span class="op">=</span> <span class="st">"voom.limma"</span>, Rmdfunction <span class="op">=</span> <span class="st">"voom.limma.createRmd"</span>, </span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>, norm.method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1.rds"</span>, </span>
<span>           result.extent <span class="op">=</span> <span class="st">"edgeR.exact"</span>, Rmdfunction <span class="op">=</span> <span class="st">"edgeR.exact.createRmd"</span>, </span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>, norm.method <span class="op">=</span> <span class="st">"TMM"</span>, </span>
<span>           trend.method <span class="op">=</span> <span class="st">"movingave"</span>, disp.type <span class="op">=</span> <span class="st">"tagwise"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"B_625_625_5spc_repl1.rds"</span>, result.extent <span class="op">=</span> <span class="st">"ttest"</span>, </span>
<span>           Rmdfunction <span class="op">=</span> <span class="st">"ttest.createRmd"</span>, </span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>, norm.method <span class="op">=</span> <span class="st">"TMM"</span><span class="op">)</span></span></code></pre></div>
<p>The code needed to perform each of the analyses is provided in the
<code>*.createRmd</code> functions. To obtain a list of all available
<code>*.createRmd</code> functions (and hence of the available
differential expression methods), we can use the
<code><a href="../reference/listcreateRmd.html">listcreateRmd()</a></code> function. Example calls are also provided
in the reference manual (see the help pages for the
<code>runDiffExp</code> function).</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listcreateRmd.html">listcreateRmd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "DESeq2.createRmd"              </span></span>
<span><span class="co">#&gt;  [2] "DESeq2.length.createRmd"       </span></span>
<span><span class="co">#&gt;  [3] "DSS.createRmd"                 </span></span>
<span><span class="co">#&gt;  [4] "EBSeq.createRmd"               </span></span>
<span><span class="co">#&gt;  [5] "edgeR.exact.createRmd"         </span></span>
<span><span class="co">#&gt;  [6] "edgeR.GLM.createRmd"           </span></span>
<span><span class="co">#&gt;  [7] "lengthNorm.limma.createRmd"    </span></span>
<span><span class="co">#&gt;  [8] "lengthNorm.sva.limma.createRmd"</span></span>
<span><span class="co">#&gt;  [9] "logcpm.limma.createRmd"        </span></span>
<span><span class="co">#&gt; [10] "NBPSeq.createRmd"              </span></span>
<span><span class="co">#&gt; [11] "NOISeq.prenorm.createRmd"      </span></span>
<span><span class="co">#&gt; [12] "phylolm.createRmd"             </span></span>
<span><span class="co">#&gt; [13] "sqrtcpm.limma.createRmd"       </span></span>
<span><span class="co">#&gt; [14] "TCC.createRmd"                 </span></span>
<span><span class="co">#&gt; [15] "ttest.createRmd"               </span></span>
<span><span class="co">#&gt; [16] "voom.limma.createRmd"          </span></span>
<span><span class="co">#&gt; [17] "voom.ttest.createRmd"</span></span></code></pre></div>
<p>You can also <a href="#providing-your-own-differential-expression-code">apply your own
differential expression method</a> to the simulated data.</p>
</div>
<div class="section level3">
<h3 id="comparing-results-from-several-differential-expression-methods">Comparing results from several differential expression methods<a class="anchor" aria-label="anchor" href="#comparing-results-from-several-differential-expression-methods"></a>
</h3>
<p>Once we have obtained the results of the differential expression
analyses (either by the methods interfaced by <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
or in other ways), we can compare the results and generate an HTML
report summarizing the results from the different methods from many
different aspects (see below for an <a href="#the-evaluation-metrics">overview of the comparison metrics</a>).
In <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>,
there are two ways of invoking the comparison functionality; either
directly from the command line or via a graphical user interface (GUI).
The GUI is mainly included to avoid long function calls and provide a
clear overview of the available methods and parameter choices. To use
the GUI the R package <em><a href="https://CRAN.R-project.org/package=rpanel" class="external-link">rpanel</a></em> must be
installed (which assumes that BWidget is available). Moreover, the GUI
may have rendering problems on certain platforms, particularly on small
screens and if many methods are to be compared. Below, we will show how
to perform the comparison using both approaches.</p>
<div class="section level4">
<h4 id="the-graphical-user-interface">The graphical user interface<a class="anchor" aria-label="anchor" href="#the-graphical-user-interface"></a>
</h4>
<p>First, we consider the <code>runComparisonGUI</code> function, to
which we provide a list of directories containing our result files, and
the directory where the final report will be generated. Since the three
result files above were saved in the current working directory, we can
run the following code to perform the comparison:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runComparisonGUI.html">runComparisonGUI</a></span><span class="op">(</span>input.directories <span class="op">=</span> <span class="st">"."</span>, </span>
<span>                 output.directory <span class="op">=</span> <span class="st">"."</span>, recursive <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<p>This opens a graphical user interface (Figure 2) where we can select
which of the data files available in the input directories that should
be included as a basis for the comparison, and which comparisons to
perform. Through this interface, we can also set p-value cutoffs for
significance and other parameters that will govern the behaviour of the
comparison.</p>
<p><strong>Important!</strong> When you have modified a value in one of
the textboxes (p-value cutoffs etc.), press Return on your keyboard to
confirm the assignment of the new value to the parameter. Always check
in the resulting comparison report that the correct values were
recognized and used for the comparisons.</p>
<div class="figure" style="text-align: center">
<img src="screenshot-gui-1.png" alt="Screenshot of the graphical user interface used to select data set (left) and set parameters (right) for the comparison of differential expression methods. The available choices for the Data set, DE methods, Number of samples and Replicates are automatically generated from the compData objects available in the designated input directories. Only one data set can be used for the comparison. In the lower part of the window we can set (adjusted) p-value thresholds for each comparison method separately. For example, we can evaluate the true FDR at one adjusted p-value threshold, and estimate the TPR for another adjusted p-value threshold. We can also set the maximal number of top-ranked variables that will be considered for the false discovery curves." width="121"><img src="screenshot-gui-2.png" alt="Screenshot of the graphical user interface used to select data set (left) and set parameters (right) for the comparison of differential expression methods. The available choices for the Data set, DE methods, Number of samples and Replicates are automatically generated from the compData objects available in the designated input directories. Only one data set can be used for the comparison. In the lower part of the window we can set (adjusted) p-value thresholds for each comparison method separately. For example, we can evaluate the true FDR at one adjusted p-value threshold, and estimate the TPR for another adjusted p-value threshold. We can also set the maximal number of top-ranked variables that will be considered for the false discovery curves." width="516"><p class="caption">
Screenshot of the graphical user interface used to select data set
(left) and set parameters (right) for the comparison of differential
expression methods. The available choices for the Data set, DE methods,
Number of samples and Replicates are automatically generated from the
compData objects available in the designated input directories. Only one
data set can be used for the comparison. In the lower part of the window
we can set (adjusted) p-value thresholds for each comparison method
separately. For example, we can evaluate the true FDR at one adjusted
p-value threshold, and estimate the TPR for another adjusted p-value
threshold. We can also set the maximal number of top-ranked variables
that will be considered for the false discovery curves.
</p>
</div>
<p>After the selections have been made, the function will perform the
comparisons and generate an HTML report, which is saved in the
designated output directory and automatically named
<code>compcodeR_report_&lt;timestamp&gt;.html</code>. Please note that
depending on the number of compared methods and the number of included
data sets, this step may take several minutes. <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
will notify you when the report is ready (with a message
<strong>Done!</strong> in the console).</p>
<p><strong>Note!</strong> Depending on the platform you use to run R,
you may see a prompt (<span class="math inline">\("&gt;"\)</span>) in the console before
the analysis is done. However, <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
will always notify you when it is finished, by typing
<strong>Done!</strong> in the console.</p>
<p>The comparison function will also generate a subdirectory called
<code>compcodeR_code</code>, where the R code used to perform each of
the comparisons is detailed in HTML reports, and another subdirectory
called <code>compcodeR_figure</code>, where the plots generated in the
comparison are saved. The code HTML reports can also be generated
manually for a given result file (given that the <code>code</code>
component is present in the <code>compData</code> object), using the
function <code>generateCodeHTMLs</code>. For example, to generate a
report containing the code that was run to perform the t-test above, as
well as the output from the R console, we can write:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/generateCodeHTMLs.html">generateCodeHTMLs</a></span><span class="op">(</span><span class="st">"B_625_625_5spc_repl1_ttest.rds"</span>, <span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<p>These reports are useful to check that there were no errors or
warnings when running the differential expression analyses.</p>
</div>
<div class="section level4">
<h4 id="direct-command-line-call">Direct command-line call<a class="anchor" aria-label="anchor" href="#direct-command-line-call"></a>
</h4>
<p>Next, we show how to run the comparison by calling the function
<code>runComparison</code> directly. In this case, we need to supply the
function with a list of result files to use as a basis for the
comparison. We can also provide a list of parameters (p-value
thresholds, the differential expression methods to include in the
comparison, etc.). The default values of these parameters are outlined
in the reference manual. The following code provides an example, given
the data generated above.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">file.table</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>input.files <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B_625_625_5spc_repl1_voom.limma.rds"</span>, </span>
<span>                                         <span class="st">"B_625_625_5spc_repl1_ttest.rds"</span>, </span>
<span>                                         <span class="st">"B_625_625_5spc_repl1_edgeR.exact.rds"</span><span class="op">)</span>, </span>
<span>                         stringsAsFactors <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>incl.nbr.samples <span class="op">=</span> <span class="cn">NULL</span>, incl.replicates <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                   incl.dataset <span class="op">=</span> <span class="st">"B_625_625"</span>, incl.de.methods <span class="op">=</span> <span class="cn">NULL</span>, </span>
<span>                   fdr.threshold <span class="op">=</span> <span class="fl">0.05</span>, tpr.threshold <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                   typeI.threshold <span class="op">=</span> <span class="fl">0.05</span>, ma.threshold <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                   fdc.maxvar <span class="op">=</span> <span class="fl">1500</span>, overlap.threshold <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                   fracsign.threshold <span class="op">=</span> <span class="fl">0.05</span>, </span>
<span>                   comparisons <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"auc"</span>, <span class="st">"fdr"</span>, <span class="st">"tpr"</span>, <span class="st">"ma"</span>, <span class="st">"correlation"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runComparison.html">runComparison</a></span><span class="op">(</span>file.table <span class="op">=</span> <span class="va">file.table</span>, parameters <span class="op">=</span> <span class="va">parameters</span>, output.directory <span class="op">=</span> <span class="st">"."</span><span class="op">)</span></span></code></pre></div>
<p>By setting <code>incl.nbr.samples</code>,
<code>incl.replicates</code> and <code>incl.de.methods</code> to
<code>NULL</code>, we ask <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
to include all results provided in the <code>file.table</code>. By
providing a vector of values for each of these variables, it is possible
to limit the selection to a subset of the provided files. Please note
that the values given to <code>incl.replicates</code> and
<code>incl.nbr.samples</code> are matched with values of the
<code>info.parameters$repl.id</code> and
<code>info.parameters$samples.per.cond</code> slots in the data/result
objects, and that the values given to <code>incl.de.methods</code> are
matched with values stored in the <code>method.names$full.name</code>
slot of the result objects. If the values do not match, the
corresponding result object will not be considered in the comparison.
Consult the package manual for the full list of comparison methods
available for use with the <code>runComparison</code> function. Setting
<code>parameters = NULL</code> implies that all results provided in the
<code>file.table</code> are used, and that all parameter values are set
to their defaults (see the reference manual). Note that only one dataset
identifier can be provided to the comparison (that is,
<code>parameters$incl.dataset</code> must be a single string).</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="using-your-own-data">Using your own data<a class="anchor" aria-label="anchor" href="#using-your-own-data"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package provides a straightforward function for simulating count data
(<code>generateSyntheticData</code>). However, it is easy to apply the
interfaced differential expression methods to your own data, given that
it is provided in a <code>compData</code> object (see <a href="#the-data-object">below</a> for a description of the data format).
You can use the function <code>check_compData</code> to check that your
object satisfies the necessary criteria to be fed into the differential
expression methods.</p>
<p>To create a <code>compData</code> object from a count matrix and a
data frame with sample annotations, you can use the function
<code>compData</code>. The following code provides a minimal example.
Note that you need to provide a <code>dataset</code> name (a
‘description’ of the simulation settings) as well as a unique data set
identifier <code>uID</code>, which has to be unique for each
<code>compData</code> object (e.g., for each simulation instance, even
if the same simulation parameters are used).</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">count.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">4000</span><span class="op">)</span><span class="op">)</span>, <span class="fl">1000</span>, <span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">sample.annot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">info.parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"mytestdata"</span>, uID <span class="op">=</span> <span class="st">"123456"</span><span class="op">)</span></span>
<span><span class="va">cpd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/compData.html">compData</a></span><span class="op">(</span>count.matrix <span class="op">=</span> <span class="va">count.matrix</span>, </span>
<span>                sample.annotations <span class="op">=</span> <span class="va">sample.annot</span>, </span>
<span>                info.parameters <span class="op">=</span> <span class="va">info.parameters</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/check_compData.html">check_compData</a></span><span class="op">(</span><span class="va">cpd</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="providing-your-own-differential-expression-code">Providing your own differential expression code<a class="anchor" aria-label="anchor" href="#providing-your-own-differential-expression-code"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package provides an interface for calling some of the most commonly used
differential expression methods developed for RNAseq data. However, it
is easy to incorporate your own favorite method. In principle, this can
be done in one of two ways:</p>
<ul>
<li>
<p>Write a <code>XXX.createRmd</code> function (where
<code>XXX</code> corresponds to your method), similar to the ones
provided in the package, which creates a <code>.Rmd</code> file
containing the code that is run to perform the differential expression
analysis. Then call this function through the <code>runDiffExp</code>
function. When implementing your function, make sure that the output is
a <code>compData</code> object, structured as described <a href="#the-result-object">below</a>. The <code>*.createRmd</code>
functions provided in the package take the following input
arguments:</p>
<ul>
<li>
<code>data.path</code> – the path to the <code>.rds</code> file
containing the <code>compData</code> object to which the differential
expression will be applied.</li>
<li>
<code>result.path</code> – the path to the <code>.rds</code> file
where the resulting <code>compData</code> object will be stored.</li>
<li>
<code>codefile</code> – the name of the code (with extension .Rmd)
where the code will be stored.</li>
<li>any arguments for setting parameters of the differential expression
analysis.</li>
</ul>
</li>
<li><p>Run the differential expression analysis completely outside the
<em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package, and save the result in a <code>compData</code> object with the
slots described <a href="#the-result-object">below</a>.</p></li>
</ul>
<p>You can use the <code>check_compData_results</code> function to check
if your object satisfies the necessary conditions for being used as the
output of a differential expression analysis and compared to results
obtained by other methods with <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>.</p>
</div>
<div class="section level2">
<h2 id="the-format-of-the-data-and-result-objects">The format of the data and result objects<a class="anchor" aria-label="anchor" href="#the-format-of-the-data-and-result-objects"></a>
</h2>
<p>This section details the format of the data and result objects
generated and used by the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package. Both objects are of the <code>compData</code> class. The format
guidelines below must be be followed if you apply the functions in the
package to a data set of your own, or to differential expression results
generated outside the package. Note that for most of the functionality
of the package, the objects should be saved separately to files with a
<code>.rds</code> extension, and the path of this object is provided to
the functions.</p>
<div class="section level3">
<h3 id="the-data-object">The data object<a class="anchor" aria-label="anchor" href="#the-data-object"></a>
</h3>
<p>The <code>compData</code> data object used by <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
is an S4 object with the following slots:</p>
<ul>
<li><p><code>count.matrix</code> [<code>class matrix</code>]
(<strong>mandatory</strong>) – the count matrix, with rows representing
genes and columns representing samples.</p></li>
<li>
<p><code>sample.annotations</code> [<code>class data.frame</code>]
(<strong>mandatory</strong>) – sample annotations. Each row corresponds
to one sample, and each column to one annotation. The data objects
generated by the <code>generateSyntheticData</code> function have two
annotations:</p>
<ul>
<li>
<code>condition</code> [<code>class character</code> or
<code>numeric</code>] (<strong>mandatory</strong>) – the class for each
sample. Currently the differential expression implementations in the
package supports only two-group comparisons, hence the
<code>condition</code> should have exactly two unique values.</li>
<li>
<code>depth.factor</code> [<code>class numeric</code>] – the depth
factor for each sample. This factor, multiplied by the
<code>sequencing depth</code>, corresponds to the target library size of
the sample when simulating the counts.</li>
</ul>
</li>
<li>
<p><code>info.parameters</code> [<code>class list</code>] – a list
detailing the parameter values that have been used for the
simulation.</p>
<ul>
<li>
<code>dataset</code> [<code>class character</code>]
(<strong>mandatory</strong>) – the name of the data set.</li>
<li>
<code>samples.per.cond</code> [<code>class numeric</code>] – the
number of samples in each of the two conditions.</li>
<li>
<code>n.diffexp</code> [<code>class numeric</code>] – the number of
genes that are simulated to be differentially expressed between the two
conditions.</li>
<li>
<code>repl.id</code> [<code>class numeric</code>] – a replicate
number, which can be set to differentiate between different instances
generated with the exact same simulation settings.</li>
<li>
<code>seqdepth</code> [<code>class numeric</code>] – the “base”
sequencing depth that was used for the simulations. For each sample, it
is modified by multiplication with a value sampled uniformly between
<code>minfact</code> and <code>maxfact</code> to generate the actual
sequencing depth for the sample.</li>
<li>
<code>minfact</code> [<code>class numeric</code>] – the lower bound
on the values used to multiply the <code>seqdepth</code> to generate the
actual sequencing depth for the individual samples.</li>
<li>
<code>maxfact</code> [<code>class numeric</code>] – the upper bound
on the values used to multiply the <code>seqdepth</code> to generate the
actual sequencing depth for the individual samples.</li>
<li>
<code>fraction.upregulated</code> [<code>class numeric</code>] – the
fraction of the differentially expressed genes that were simulated to be
upregulated in condition 2 compared to condition 1. Must be in the
interval [0, 1]. The remaining differentially expressed genes are
simulated to be downregulated in condition 2.</li>
<li>
<code>between.group.diffdisp</code> [<code>class logical</code>] –
whether the counts from the two conditions were simulated with different
dispersion parameters.</li>
<li>
<code>filter.threshold.total</code> [<code>class numeric</code>] –
the filter threshold that is applied to the total count across all
samples. All genes for which the total count is below this number have
been excluded from the data set in the simulation process.</li>
<li>
<code>filter.threshold.mediancpm</code> [<code>class numeric</code>]
– the filter threshold that is applied to the median count per million
(cpm) across all samples. All genes for which the median cpm is below
this number have been excluded from the data set in the simulation
process.</li>
<li>
<code>fraction.non.overdispersed</code> [<code>class numeric</code>]
– the fraction of the genes in the data set that are simulated according
to a model without overdispersion (i.e., a Poisson model).</li>
<li>
<code>random.outlier.high.prob</code> [<code>class numeric</code>] –
the fraction of extremely high random outlier counts introduced in the
simulated data sets. Please consult <span class="citation">Soneson and
Delorenzi (2013)</span> for a detailed description of the different
types of outliers.</li>
<li>
<code>random.outlier.low.prob</code> [<code>class numeric</code>] –
the fraction of extremely low random outlier counts introduced in the
simulated data sets. Please consult <span class="citation">Soneson and
Delorenzi (2013)</span> for a detailed description of the different
types of outliers.</li>
<li>
<code>single.outlier.high.prob</code> [<code>class numeric</code>] –
the fraction of extremely high single outlier counts introduced in the
simulated data sets. Please consult <span class="citation">Soneson and
Delorenzi (2013)</span> for a detailed description of the different
types of outliers.</li>
<li>
<code>single.outlier.low.prob</code> [<code>class numeric</code>] –
the fraction of extremely low single outlier counts introduced in the
simulated data sets. Please consult <span class="citation">Soneson and
Delorenzi (2013)</span> for a detailed description of the different
types of outliers.</li>
<li>
<code>effect.size</code> [<code>class numeric</code>] – the degree
of differential expression, i.e., a measure of the minimal effect size.
Alternatively, a vector of provided effect sizes for each of the
variables.</li>
<li>
<code>uID</code> [<code>class character</code>]
(<strong>mandatory</strong>) – a unique identification number given to
each data set. For the data sets simulated with <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>,
it consists of a sequence of 10 randomly generated alpha-numeric
characters.</li>
</ul>
</li>
<li><p><code>filtering</code> [<code>class character</code>] – a summary
of the filtering that has been applied to the data.</p></li>
<li>
<p><code>variable.annotations</code> [<code>class data.frame</code>]
– annotations for each of the variables in the data set. Each row
corresponds to one variable, and each column to one annotation. No
annotation is mandatory, however some of them are necessary if the
differential expression results for the data are going to be used for
certain comparison tasks. The data sets simulated within <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
have the following named variable annotations:</p>
<ul>
<li>
<code>truedispersions.S1</code> [<code>class numeric</code>] – the
true dispersions used in the simulations of the counts for the samples
in condition 1.</li>
<li>
<code>truedispersions.S2</code> [<code>class numeric</code>] – the
true dispersions used in the simulations of the counts for the samples
in condition 2.</li>
<li>
<code>truemeans.S1</code> [<code>class numeric</code>] – the true
mean values used in the simulations of the counts for the samples in
condition 1.</li>
<li>
<code>truemeans.S2</code> [<code>class numeric</code>] – the true
mean values used in the simulations of the counts for the samples in
condition 2.</li>
<li>
<code>n.random.outliers.up.S1</code> [<code>class numeric</code>] –
the number of extremely high random outliers introduced for each gene in
condition 1.</li>
<li>
<code>n.random.outliers.up.S2</code> [<code>class numeric</code>] –
the number of extremely high random outliers introduced for each gene in
condition 2.</li>
<li>
<code>n.random.outliers.down.S1</code> [<code>class numeric</code>]
– the number of extremely low random outliers introduced for each gene
in condition 1.</li>
<li>
<code>n.random.outliers.down.S2</code> [<code>class numeric</code>]
– the number of extremely low random outliers introduced for each gene
in condition 2.</li>
<li>
<code>n.single.outliers.up.S1</code> [<code>class numeric</code>] –
the number of extremely high single outliers introduced for each gene in
condition 1.</li>
<li>
<code>n.single.outliers.up.S2</code> [<code>class numeric</code>] –
the number of extremely high single outliers introduced for each gene in
condition 2.</li>
<li>
<code>n.single.outliers.down.S1</code> [<code>class numeric</code>]
– the number of extremely low single outliers introduced for each gene
in condition 1.</li>
<li>
<code>n.single.outliers.down.S2</code> [<code>class numeric</code>]
– the number of extremely low single outliers introduced for each gene
in condition 2.</li>
<li>
<code>M.value</code> [<code>class numeric</code>] – the estimated
log2-fold change between conditions 1 and 2 for each gene. These values
were estimated using the <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>
package.</li>
<li>
<code>A.value</code> [<code>class numeric</code>] – the estimated
average expression in conditions 1 and 2 for each gene. These values
were estimated using the <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>
package.</li>
<li>
<code>truelog2foldchanges</code> [<code>class numeric</code>] – the
“true” log2-fold changes between conditions 1 and 2 for each gene, based
on the parameters used for simulation.</li>
<li>
<code>upregulation</code> [<code>class numeric</code>] – a binary
annotation (0/1) indicating which genes are simulated to be upregulated
in condition 2 compared to condition 1. (Upregulated genes are indicated
with a 1.)</li>
<li>
<code>downregulation</code> [<code>class numeric</code>] – a binary
annotation (0/1) indicating which genes are simulated to be
downregulated in condition 2 compared to condition 1. (Downregulated
genes are indicated with a 1.)</li>
<li>
<code>differential.expression</code> [<code>class numeric</code>]
(mandatory for many comparisons, such as the computation of false
discovery rates, true positive rates, ROC curves, false discovery curves
etc.)– a binary annotation (0/1) indicating which genes are simulated to
be differentially expressed in condition 2 compared to condition 1. In
other words, the sum of <code>upregulation</code> and
<code>downregulation</code>.</li>
</ul>
</li>
</ul>
<p>To apply the functions of the package to a <code>compData</code>
object of the type detailed above, it needs to be saved to a file with
extension <code>.rds</code>. To save the object <code>cpd</code> to the
file <code>saveddata.rds</code>, simply type</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="va">cpd</span>, <span class="st">"saveddata.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="the-result-object">The result object<a class="anchor" aria-label="anchor" href="#the-result-object"></a>
</h3>
<p>When applying one of the differential expression methods interfaced
through <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>,
the <code>compData</code> object is extended with some additional slots.
These are described below.</p>
<ul>
<li><p><code>analysis.date</code> [<code>class character</code>] – the
date and time when the differential expression analysis was
performed.</p></li>
<li><p><code>package.version</code> [<code>class character</code>] – the
version of the packages used in the differential expression
analysis.</p></li>
<li>
<p><code>method.names</code> [<code>class list</code>]
(<strong>mandatory</strong>) – a list containing the names of the
differential expression method, which should be used to identify the
results in the comparison. Contains two components:</p>
<ul>
<li>
<code>short.name</code> [<code>class character</code>] – a short
name, used for convenience.</li>
<li>
<code>full.name</code> [<code>class character</code>] – a fully
identifying name of the differential expression method, including
e.g. version numbers and parameter values. In the method comparisons,
the results will be grouped based on the <code>full.name</code>.</li>
</ul>
</li>
<li><p><code>code</code> [<code>class character</code>] – the code that
was used to run the differential expression analysis, in R markdown
(.Rmd) format.</p></li>
<li>
<p><code>result.table</code> [<code>class data.frame</code>]
(<strong>mandatory</strong>) – a data frame containing the results of
the differential expression analysis. Each row corresponds to one gene,
and each column to a value generated by the analysis. The precise
columns will depend on the method applied. The following columns are
used by at least one of the methods interfaced by <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>:</p>
<ul>
<li>
<code>pvalue</code> [<code>class numeric</code>] – the nominal
p-values.</li>
<li>
<code>adjpvalue</code> [<code>class numeric</code>] – p-values
adjusted for multiple comparisons.</li>
<li>
<code>logFC</code> [<code>class numeric</code>] – estimated log-fold
changes between the two conditions.</li>
<li>
<code>score</code> [<code>class numeric</code>]
(<strong>mandatory</strong>) – the score that will be used to rank the
genes in order of significance. Note that <strong>high scores always
signify differential expression, that is, a strong association with the
predictor</strong>. For example, for methods returning a nominal p-value
the score is generally obtained as <code>1 - pvalue</code>.</li>
<li>
<code>FDR</code> [<code>class numeric</code>] – the false discovery
rate estimate.</li>
<li>
<code>posterior.DE</code> [<code>class numeric</code>] – the
posterior probability of differential expression.</li>
<li>
<code>prob.DE</code> [<code>class numeric</code>] – The conditional
probability of differential expression.</li>
<li>
<code>lfdr</code> [<code>class numeric</code>] – the local false
discovery rate.</li>
<li>
<code>statistic</code> [<code>class numeric</code>] – a test
statistic from the differential expression analysis.</li>
<li>
<code>dispersion.S1</code> [<code>class numeric</code>] – dispersion
estimate in condition 1.</li>
<li>
<code>dispersion.S2</code> [<code>class numeric</code>] – dispersion
estimate in condition 2.</li>
</ul>
</li>
</ul>
<p>For many of the comparison methods, the naming of the result columns
is important. For example, the p-value column must be named
<code>pvalue</code> in order to be recognized by the comparison method
computing type I error. Similarly, either an <code>adjpvalue</code> or
an <code>FDR</code> column must be present in order to apply the
comparison methods requiring adjusted p-value/FDR cutoffs. If both are
present, the <code>adjpvalue</code> column takes precedence over the
<code>FDR</code> column.</p>
<p>To be used in the comparison function, the result
<code>compData</code> object must be saved to a <code>.rds</code>
file.</p>
</div>
</div>
<div class="section level2">
<h2 id="the-evaluation-metrics">The evaluation metrics<a class="anchor" aria-label="anchor" href="#the-evaluation-metrics"></a>
</h2>
<p>This section provides an overview of the methods that are implemented
in <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
for comparing differential expression results obtained by different
methods. The selection of which methods to apply is made through a
graphical user interface that is opened when the
<code>runComparisonGUI</code> function is called (see Figure 2).
Alternatively, the selection of methods can be supplied to the
<code>runComparison</code> function directly, to circumvent the GUI.</p>
<div class="section level3">
<h3 id="roc-one-replicateall-replicates">ROC (one replicate/all replicates)<a class="anchor" aria-label="anchor" href="#roc-one-replicateall-replicates"></a>
</h3>
<p>This method computes a ROC curve for either a single representative
among the replicates of a given data set, or for all replicates (in
separate plots). The ROC curves are generated by plotting the true
positive rate (TPR, on the y-axis) versus the false positive rate (FPR,
on the x-axis) when varying the cutoff on the <code>score</code> (see
the description of the result object above). A good ranking method gives
a ROC curve which passes close to the upper left corner of the plot,
while a bad method gives a ROC curve closer to the diagonal. Calculation
of the ROC curves requires that the differential expression status of
each gene is provided.</p>
<p><img src="compcodeR_figure/rocone-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="auc">AUC<a class="anchor" aria-label="anchor" href="#auc"></a>
</h3>
<p>This comparison method computes the area under the ROC curve and
represents the result in boxplots, where each box summarizes the results
for one method across all replicates of a data set. A good ranking
method gives to a large value of the AUC. This requires that the
differential expression status of each gene is provided.</p>
<p><img src="compcodeR_figure/auc.png" width="504" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="type-i-error">Type I error<a class="anchor" aria-label="anchor" href="#type-i-error"></a>
</h3>
<p>This approach computes the type I error (the fraction of the genes
that are truly non-differentially expressed that are called significant
at a given nominal p-value threshold) for each method and sample size
separately, and represent it in boxplots, where each box summarizes the
results for one method across all replicates of a data set. This
requires that the differential expression status of each gene is
provided (see the description of the data object above) and that the
<code>pvalue</code> column is present in the <code>result.table</code>
of the included result objects. The dashed vertical line represents the
imposed nominal p-value threshold, and we wish that the observed type I
error is lower than this threshold.</p>
<p><img src="compcodeR_figure/typeIerror.png" width="504" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="fdr">FDR<a class="anchor" aria-label="anchor" href="#fdr"></a>
</h3>
<p>Here, we compute the observed false discovery rate (the fraction of
the genes called significant that are truly non-differentially expressed
at a given adjusted p-value/FDR threshold) for each method and sample
size separately, and represent the estimates in boxplots, where each box
summarizes the results for one method across all replicates of a data
set. This requires that the differential expression status of each gene
is provided (see the description of the data object above) and that the
<code>adjpvalue</code> or <code>FDR</code> column is present in the
<code>result.table</code> of the included result objects. The dashed
vertical line represents the imposed adjusted p-value threshold (that
is, the level at which we wish to control the false discovery rate).</p>
<p><img src="compcodeR_figure/fdr.png" width="504" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="fdr-as-a-function-of-expression-level">FDR as a function of expression level<a class="anchor" aria-label="anchor" href="#fdr-as-a-function-of-expression-level"></a>
</h3>
<p>Instead of looking at the overall false discovery rate, this method
allows us to study the FDR as a function of the average expression level
of the genes. For each data set, the average expression levels are
binned into 10 bins of equal size (i.e., each containing 10% of the
genes), and the FDR is computed for each of them. The results are shown
by means of boxplots, summarizing the results across all replicates of a
data set.</p>
<p><img src="compcodeR_figure/fdrvsexpr-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="tpr">TPR<a class="anchor" aria-label="anchor" href="#tpr"></a>
</h3>
<p>With this approach, we compute the observed true positive rate (the
fraction of the truly differentially expressed genes that are called
significant at a given adjusted p-value/FDR threshold) for each method
and sample size separately, and represent the estimates in boxplots,
where each box summarizes the results for one method across all
replicates of a data set. This requires that the differential expression
status of each gene is provided (see the description of the data object
above) and that the <code>adjpvalue</code> or <code>FDR</code> column is
present in the <code>result.table</code> of the included result objects.
There is often a trade-off between achieving a high TPR (which is
desirable) and controlling the number of false positives, and hence the
TPR plots should typically be interpreted together with the type I error
and/or FDR plots.</p>
<p><img src="compcodeR_figure/tpr.png" width="504" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="false-discovery-curves-one-replicateall-replicates">False discovery curves (one replicate/all replicates)<a class="anchor" aria-label="anchor" href="#false-discovery-curves-one-replicateall-replicates"></a>
</h3>
<p>This choice plots the false discovery curves, depicting the number of
false discoveries (i.e., truly non-differentially expressed genes) that
are encountered when traversing the ranked list of genes ordered in
decreasing order by the <code>score</code> column of the
<code>result.table</code>. As for the ROC curves, the plots can be made
for a single replicate or for all replicates (each in a separate
figure). A well-performing method is represented by a slowly rising
false discovery curve (in other words, few false positives among the
top-ranked genes).</p>
<p><img src="compcodeR_figure/fdcone-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="fractionnumber-of-significant-genes">Fraction/number of significant genes<a class="anchor" aria-label="anchor" href="#fractionnumber-of-significant-genes"></a>
</h3>
<p>We can also compute the fraction and/or number of the genes that are
called significant at a given adjusted p-value/FDR threshold for each
method and sample size separately, and represent the estimates in
boxplots, where each box summarizes the results for one method across
all replicates of a data set. This requires that the
<code>adjpvalue</code> or <code>FDR</code> column is present in the
<code>result.table</code> of the included result objects. Since this
plot does not incorporate information about the number of the identity
of the truly differentially expressed genes, we can not conclude whether
a large or small fraction of significant genes is preferable. Thus, the
plot is merely an indication of which methods are more liberal (giving
more significantly differentially expressed genes) and which are more
conservative.</p>
<p><img src="compcodeR_figure/fracsign.png" width="504" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="overlap-one-replicate">Overlap, one replicate<a class="anchor" aria-label="anchor" href="#overlap-one-replicate"></a>
</h3>
<p>For each pair of methods, this approach computes the overlap between
the sets of genes called differentially expressed by each of them at a
given adjusted p-value/FDR threshold. Only one representative of the
replicates of a data set is used. This requires that the
<code>adjpvalue</code> or <code>FDR</code> column is present in the
<code>result.table</code> of the included result objects. Since the
overlap depends on the number of genes called differentially expressed
by the different methods, a normalized overlap measure (the Sorensen
index) is also provided (see below). The results are represented as a
table of the sizes of the overlaps between each pair of methods.</p>
</div>
<div class="section level3">
<h3 id="sorensen-index-one-replicate">Sorensen index, one replicate<a class="anchor" aria-label="anchor" href="#sorensen-index-one-replicate"></a>
</h3>
<p>For each pair of methods, this approach computes a normalized
pairwise overlap value between the sets of genes called differentially
expressed by each of them at a given adjusted p-value/FDR threshold.
Only one representative of the replicates of a data set is used. This
requires that the <code>adjpvalue</code> or <code>FDR</code> column is
present in the <code>result.table</code> of the included result objects.
The Sorensen index (also called the Dice coefficient) for two sets <span class="math inline">\(A\)</span> and <span class="math inline">\(B\)</span> is given by <span class="math display">\[S(A,B)=\frac{2|A\cap B|}{|A|+|B|},\]</span>where
<span class="math inline">\(|\cdot|\)</span> denotes the cardinality of
a set (the number of elements in the set). The results are represented
in the form of a table, and also as a colored heatmap, where the color
represents the degree of overlap between the sets of differentially
expressed genes found by different methods.</p>
<p><img src="compcodeR_figure/sorensen-1.png" width="216" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="ma-plot">MA plot<a class="anchor" aria-label="anchor" href="#ma-plot"></a>
</h3>
<p>With this approach we can construct MA-plots (one for each
differential expression method) for one replicate of a data set,
depicting the estimated log-fold change (on the y-axis) versus the
estimated average expression (on the x-axis). The genes that are called
differentially expressed are marked with color.</p>
<p><img src="compcodeR_figure/maplot-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="spearman-correlation">Spearman correlation<a class="anchor" aria-label="anchor" href="#spearman-correlation"></a>
</h3>
<p>This approach computes the pairwise Spearman correlation between the
gene scores obtained by different differential expression methods (the
<code>score</code> component of the <code>result.table</code>). The
correlations are visualized by means of a color-coded table and used to
construct a dissimilarity measure for hierarchical clustering of the
differential expression methods.</p>
<p><img src="compcodeR_figure/correlation-11.png" width="216" style="display: block; margin: auto;"><img src="compcodeR_figure/correlation-12.png" width="216" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="score-distribution-vs-number-of-outliers">Score distribution vs number of outliers<a class="anchor" aria-label="anchor" href="#score-distribution-vs-number-of-outliers"></a>
</h3>
<p>This method plots the distribution of the gene scores obtained by
different differential expression methods (the <code>score</code>
component of the <code>result.table</code>, see the description of the
result object above) as a function of the number of outliers imposed for
the genes. The distributions are visualized by means of violin plots. In
the example below, there are no outliers introduced. Recall that a high
score corresponds to more significant genes. In the presence of
outliers, the score distribution plots can be used to examine whether
the presence of an outlier count for a gene shifts the score
distribution towards higher or lower values.</p>
<p><img src="compcodeR_figure/scorevsoutlier-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="score-distribution-vs-average-expression-level">Score distribution vs average expression level<a class="anchor" aria-label="anchor" href="#score-distribution-vs-average-expression-level"></a>
</h3>
<p>This method plots the gene scores (the <code>score</code> component
of the <code>result.table</code>, see the description of the result
object above) as a function of the average expression level of the gene.
Recall that high scores correspond to more significant genes. The
colored line shows the trend in the relationship between the two
variables by means of a loess fit. These plots can be used to examine
whether highly expressed genes tend to have (e.g.) higher scores than
lowly expressed genes, and more generally to provide a useful
characteristic of the methods. However, it is not clear what would be
the “optimal” behavior.</p>
<p><img src="compcodeR_figure/scorevsexpr-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="score-vs-signal-for-genes-expressed-in-only-one-condition">Score vs ‘signal’ for genes expressed in only one condition<a class="anchor" aria-label="anchor" href="#score-vs-signal-for-genes-expressed-in-only-one-condition"></a>
</h3>
<p>This method plots the gene scores (the <code>score</code> component
of the <code>result.table</code>, see the description of the result
object above) as a function of the ‘signal strength’ for genes that are
detected in only one of the two conditions. The signal strength can be
defined either as the average (log-transformed) normalized pseudo-count
in the condition where the gene is expressed, or as the signal-to-noise
ratio in this condition, that is, as the average log-transformed
normalized pseudo-count divided by the standard deviation of the
log-transformed normalized pseudo-count. Typically, we would expect that
the score increases with the signal strength.</p>
<p><img src="compcodeR_figure/scorevssignal-1.png" width="648" style="display: block; margin: auto;"></p>
</div>
<div class="section level3">
<h3 id="matthews-correlation-coefficient">Matthew’s correlation coefficient<a class="anchor" aria-label="anchor" href="#matthews-correlation-coefficient"></a>
</h3>
<p>Matthew’s correlation coefficient summarizes the result of a
classification task in a single number, incorporating the number of
false positives, true positives, false negatives and true negatives. A
correlation coefficient of +1 indicates perfect classification, while a
correlation coefficient of -1 indicates perfect “anti-classification”,
i.e., that all objects are misclassified.</p>
</div>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.7</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [6] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] compcodeR_1.37.4 sm_2.2-5.7.1     BiocStyle_2.29.2</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] tidyselect_1.2.0     timeDate_4022.108   </span></span>
<span><span class="co">#&gt;  [3] dplyr_1.1.3          bitops_1.0-7        </span></span>
<span><span class="co">#&gt;  [5] fastmap_1.1.1        promises_1.2.1      </span></span>
<span><span class="co">#&gt;  [7] digest_0.6.33        rpart_4.1.19        </span></span>
<span><span class="co">#&gt;  [9] mime_0.12            lifecycle_1.0.3     </span></span>
<span><span class="co">#&gt; [11] cluster_2.1.4        ellipsis_0.3.2      </span></span>
<span><span class="co">#&gt; [13] statmod_1.5.0        ROCR_1.0-11         </span></span>
<span><span class="co">#&gt; [15] magrittr_2.0.3       compiler_4.3.1      </span></span>
<span><span class="co">#&gt; [17] rlang_1.1.1          sass_0.4.7          </span></span>
<span><span class="co">#&gt; [19] tools_4.3.1          utf8_1.2.3          </span></span>
<span><span class="co">#&gt; [21] yaml_2.3.7           knitr_1.44          </span></span>
<span><span class="co">#&gt; [23] statip_0.2.3         KernSmooth_2.23-22  </span></span>
<span><span class="co">#&gt; [25] purrr_1.0.2          timeSeries_4031.107 </span></span>
<span><span class="co">#&gt; [27] fBasics_4022.94      desc_1.4.2          </span></span>
<span><span class="co">#&gt; [29] grid_4.3.1           fansi_1.0.4         </span></span>
<span><span class="co">#&gt; [31] caTools_1.18.2       stabledist_0.7-1    </span></span>
<span><span class="co">#&gt; [33] xtable_1.8-4         colorspace_2.1-0    </span></span>
<span><span class="co">#&gt; [35] edgeR_3.99.0         ggplot2_3.4.3       </span></span>
<span><span class="co">#&gt; [37] scales_1.2.1         gtools_3.9.4        </span></span>
<span><span class="co">#&gt; [39] MASS_7.3-60          cli_3.6.1           </span></span>
<span><span class="co">#&gt; [41] rmarkdown_2.25       ragg_1.2.5          </span></span>
<span><span class="co">#&gt; [43] generics_0.1.3       cachem_1.0.8        </span></span>
<span><span class="co">#&gt; [45] stringr_1.5.0        stable_1.1.6        </span></span>
<span><span class="co">#&gt; [47] BiocManager_1.30.22  vctrs_0.6.3         </span></span>
<span><span class="co">#&gt; [49] jsonlite_1.8.7       bookdown_0.35       </span></span>
<span><span class="co">#&gt; [51] rmutil_1.1.10        clue_0.3-65         </span></span>
<span><span class="co">#&gt; [53] systemfonts_1.0.4    locfit_1.5-9.8      </span></span>
<span><span class="co">#&gt; [55] limma_3.57.8         jquerylib_0.1.4     </span></span>
<span><span class="co">#&gt; [57] spatial_7.3-17       glue_1.6.2          </span></span>
<span><span class="co">#&gt; [59] pkgdown_2.0.7.9000   stringi_1.7.12      </span></span>
<span><span class="co">#&gt; [61] gtable_0.3.4         later_1.3.1         </span></span>
<span><span class="co">#&gt; [63] shinydashboard_0.7.2 munsell_0.5.0       </span></span>
<span><span class="co">#&gt; [65] tibble_3.2.1         pillar_1.9.0        </span></span>
<span><span class="co">#&gt; [67] htmltools_0.5.6      gplots_3.1.3        </span></span>
<span><span class="co">#&gt; [69] R6_2.5.1             textshaping_0.3.6   </span></span>
<span><span class="co">#&gt; [71] rprojroot_2.0.3      vioplot_0.4.0       </span></span>
<span><span class="co">#&gt; [73] evaluate_0.21        shiny_1.7.5         </span></span>
<span><span class="co">#&gt; [75] lattice_0.21-8       markdown_1.8        </span></span>
<span><span class="co">#&gt; [77] png_0.1-8            modeest_2.4.0       </span></span>
<span><span class="co">#&gt; [79] memoise_2.0.1        httpuv_1.6.11       </span></span>
<span><span class="co">#&gt; [81] bslib_0.5.1          Rcpp_1.0.11         </span></span>
<span><span class="co">#&gt; [83] xfun_0.40            fs_1.6.3            </span></span>
<span><span class="co">#&gt; [85] zoo_1.8-12           pkgconfig_2.0.3</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Soneson2013" class="csl-entry">
Soneson, Charlotte, and Mauro Delorenzi. 2013. <span>“A Comparison of
Methods for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.”</span> <em>BMC Bioinformatics</em>
14: 91.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Charlotte Soneson, Paul Bastide, Mélina Gallopin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
