<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Including inter-species measurements in differential expression analysis of RNAseq data with the compcodeR package • compcodeR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Including inter-species measurements in differential expression analysis of RNAseq data with the compcodeR package">
<meta property="og:description" content="compcodeR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">compcodeR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.37.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/compcodeR.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/phylocompcodeR.html">Including inter-species measurements in differential expression analysis of RNAseq data with the compcodeR package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/csoneson/compcodeR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Including inter-species measurements in
differential expression analysis of RNAseq data with the compcodeR
package</h1>
                        <h4 data-toc-skip class="author">Paul Bastide
&amp; Mélina Gallopin</h4>
            
            <h4 data-toc-skip class="date">2023-06-25</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/csoneson/compcodeR/blob/HEAD/vignettes/phylocompcodeR.Rmd" class="external-link"><code>vignettes/phylocompcodeR.Rmd</code></a></small>
      <div class="hidden name"><code>phylocompcodeR.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/csoneson/compcodeR" class="external-link">compcodeR</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: sm</span></span>
<span><span class="co">#&gt; Package 'sm', version 2.2-5.7: type help(sm) for summary information</span></span>
<span><span class="co">#&gt; Warning in fun(libname, pkgname): no display name and</span></span>
<span><span class="co">#&gt; no $DISPLAY environment variable</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
R package can generate RNAseq counts data and compare the relative
performances of various popular differential analysis detection tools
(<span class="citation">Soneson and Delorenzi (2013)</span>).</p>
<p>Using the same framework, this document shows how to generate
“orthologous gene” (OG) expression for different species, taking into
account their varying lengths, and their phylogenetic relationships, as
encoded by an evolutionary tree.</p>
<p>This vignette provides a tutorial on how to use the “phylogenetic”
functionalities of <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>.
It assumes that the reader is already familiar with the <a href="compcodeR.html"><code>compcodeR</code> package vignette</a>.</p>
</div>
<div class="section level2">
<h2 id="the-phylocompdata-class">The <code>phyloCompData</code> class<a class="anchor" aria-label="anchor" href="#the-phylocompdata-class"></a>
</h2>
<p>The <code>phyloCompData</code> class extends the
<code>compData</code> class of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package to account for phylogeny and length information needed in the
representation of OG expression data.</p>
<p>A <code>phyloCompData</code> object contains all the slots of a <a href="compcodeR.html#the-compdata-class"><code>compData</code>
object</a>, with an added slot containing a phylogenetic tree with <a href="https://CRAN.R-project.org/package=ape" class="external-link"><code>ape</code></a>
format <code>phylo</code>, and a length matrix. It can also contain some
added variable information, such as species names. More detailed
information about the <code>phyloCompData</code> class are available in
the section on <a href="#the-extended-data-object">the phylo data
object</a>. After conducting a differential expression analysis, the
<code>phyloCompData</code> object has the same added information than
the <code>compData</code> object (see <a href="compcodeR.html#the-result-object">the result object</a> in the
<em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package vignette).</p>
</div>
<div class="section level2">
<h2 id="a-sample-workflow">A sample workflow<a class="anchor" aria-label="anchor" href="#a-sample-workflow"></a>
</h2>
<p>The workflow for working with the inter-species extension is very
similar to the <a href="compcodeR.html#a-sample-workflow">already
existing workflow</a> of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
package. In this section, we recall this workflow, stressing out the
added functionalities.</p>
<div class="section level3">
<h3 id="phylogenetic-tree">Phylogenetic Tree<a class="anchor" aria-label="anchor" href="#phylogenetic-tree"></a>
</h3>
<p>The simulations are performed following the description by <span class="citation">Bastide et al. (2022)</span>.</p>
<p>We use here the phylogenetic tree issued from <span class="citation">Stern et al. (2017)</span>, normalized to unit height,
that has <span class="math inline">\(14\)</span> species with up to 3
replicates, for a total number of sample equal to <span class="math inline">\(34\)</span> (see Figure below).</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://ape-package.ird.fr/" class="external-link">ape</a></span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"Stern2018.tree"</span>, package <span class="op">=</span> <span class="st">"compcodeR"</span><span class="op">)</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span></code></pre></div>
<p>Note that any other tree could be used, for instance randomly
generated using a birth-death process, see e.g. function
<code>rphylo</code> in the <a href="https://CRAN.R-project.org/package=ape" class="external-link"><code>ape</code></a>
package.</p>
</div>
<div class="section level3">
<h3 id="condition-design">Condition Design<a class="anchor" aria-label="anchor" href="#condition-design"></a>
</h3>
<p>To conduct a differential analysis, each species must be attributed a
condition. Because of the phylogenetic structure, the condition design
does matter, and have a strong influence on the data produced. Here, we
assume that the conditions are mapped on the tree in a balanced way
(“alt” design), which is the “best case scenario”.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># link each sample to a species</span></span>
<span><span class="va">id_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"_.*"</span>, <span class="st">""</span>, <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">id_species</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="co"># Assign a condition to each species</span></span>
<span><span class="va">species_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">id_species</span><span class="op">)</span></span>
<span><span class="va">species_names</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">species_names</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span><span class="op">-</span><span class="fl">1</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">cond_species</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">species_names</span><span class="op">)</span> <span class="op">/</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cond_species</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">species_names</span></span>
<span><span class="co"># map them on the tree</span></span>
<span><span class="va">id_cond</span> <span class="op">&lt;-</span> <span class="va">id_species</span></span>
<span><span class="va">id_cond</span> <span class="op">&lt;-</span> <span class="va">cond_species</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="va">id_cond</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">id_cond</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">id_cond</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">id_cond</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span></code></pre></div>
<p>We can plot the assigned conditions on the tree to visualize
them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">tree</span>, label.offset <span class="op">=</span> <span class="fl">0.01</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/nodelabels.html" class="external-link">tiplabels</a></span><span class="op">(</span>pch <span class="op">=</span> <span class="fl">19</span>, col <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#D55E00"</span>, <span class="st">"#009E73"</span><span class="op">)</span><span class="op">[</span><span class="va">id_cond</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<img src="phylocompcodeR_files/figure-html/unnamed-chunk-2-1.png" alt="Phylogenetic tree with $14$ species and $34$ samples, with two conditions" width="700"><p class="caption">
Phylogenetic tree with <span class="math inline">\(14\)</span> species
and <span class="math inline">\(34\)</span> samples, with two conditions
</p>
</div>
</div>
<div class="section level3">
<h3 id="simulating-data">Simulating data<a class="anchor" aria-label="anchor" href="#simulating-data"></a>
</h3>
<p>Using this tree with associated condition design, we can then
generate a dataset using a “phylogenetic Poisson Log Normal” (pPLN)
distribution. We use here a Brownian Motion (BM) model of evolution for
the latent phylogenetic log normal continuous trait, and assume that the
phylogenetic model accounts for <span class="math inline">\(90\%\)</span> of the latent trait variance
(i.e. there is an added uniform intra-species variance representing
<span class="math inline">\(10\%\)</span> of the total latent trait
variation). Using the <code>"auto"</code> setup, the counts are
simulated so that they match empirical moments found in <span class="citation">Stern and Crandall (2018)</span>. OG lengths are also
drawn from a pPLN model, so that their moments match those of the
empirical dataset of <span class="citation">Stern and Crandall
(2018)</span>. We choose to simulate <span class="math inline">\(2000\)</span> OGs, <span class="math inline">\(10\%\)</span> of which are differentially
expressed, with an effect size of <span class="math inline">\(3\)</span>.</p>
<p>The following code creates a <code>phyloCompData</code> object
containing the simulated data set and saves it to a file named
<code>"alt_BM_repl1.rds"</code>.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">12890926</span><span class="op">)</span></span>
<span><span class="va">alt_BM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/generateSyntheticData.html">generateSyntheticData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"alt_BM"</span>,</span>
<span>                                n.vars <span class="op">=</span> <span class="fl">2000</span>, samples.per.cond <span class="op">=</span> <span class="fl">17</span>,</span>
<span>                                n.diffexp <span class="op">=</span> <span class="fl">200</span>, repl.id <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                                seqdepth <span class="op">=</span> <span class="fl">1e7</span>, effect.size <span class="op">=</span> <span class="fl">3</span>,</span>
<span>                                fraction.upregulated <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>                                output.file <span class="op">=</span> <span class="st">"alt_BM_repl1.rds"</span>,</span>
<span>                                <span class="co">## Phylogenetic parameters</span></span>
<span>                                tree <span class="op">=</span> <span class="va">tree</span>,                      <span class="co">## Phylogenetic tree</span></span>
<span>                                id.species <span class="op">=</span> <span class="va">id_species</span>,          <span class="co">## Species structure of samples</span></span>
<span>                                id.condition <span class="op">=</span> <span class="va">id_cond</span>,           <span class="co">## Condition design</span></span>
<span>                                model.process <span class="op">=</span> <span class="st">"BM"</span>,             <span class="co">## The latent trait follows a BM</span></span>
<span>                                prop.var.tree <span class="op">=</span> <span class="fl">0.9</span>,              <span class="co">## Tree accounts for 90% of the variance</span></span>
<span>                                lengths.relmeans <span class="op">=</span> <span class="st">"auto"</span>,        <span class="co">## OG length mean and dispersion</span></span>
<span>                                lengths.dispersions <span class="op">=</span> <span class="st">"auto"</span><span class="op">)</span>     <span class="co">## are taken from an empirical exemple</span></span></code></pre></div>
<p>The <code>summarizeSyntheticDataSet</code> works the same way as in
the base <a href="compcodeR.html"><code>compcodeR</code> package</a>,
generating a report that summarize all the parameters used in the
simulation, and showing some diagnostic plots.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/summarizeSyntheticDataSet.html">summarizeSyntheticDataSet</a></span><span class="op">(</span>data.set <span class="op">=</span> <span class="st">"alt_BM_repl1.rds"</span>, </span>
<span>                          output.filename <span class="op">=</span> <span class="st">"alt_BM_repl1_datacheck.html"</span><span class="op">)</span></span></code></pre></div>
<p>When applied to a <code>phyloCompData</code> object, it provides some
extra diagnostics, related to the phylogenetic nature of the data. In
particular, it contains MA-plots with TPM-normalized expression levels
to take OG length into account, which generally makes the original
signal clearer.</p>
<div class="figure" style="text-align: center">
<img src="phylocompcodeR_check_figure/maplot-trueDEstatus-1.png" alt="Example figures from the summarization report generated for a simulated data set. The top panel shows an MA plot, with the genes colored by the true differential expression status. The bottom panel shows the same plot, but using TPM-normalized estimated expression levels." width="324"><img src="phylocompcodeR_check_figure/maplot-trueDEstatus-logTPM-1.png" alt="Example figures from the summarization report generated for a simulated data set. The top panel shows an MA plot, with the genes colored by the true differential expression status. The bottom panel shows the same plot, but using TPM-normalized estimated expression levels." width="324"><p class="caption">
Example figures from the summarization report generated for a simulated
data set. The top panel shows an MA plot, with the genes colored by the
true differential expression status. The bottom panel shows the same
plot, but using TPM-normalized estimated expression levels.
</p>
</div>
<p>It also shows a log2 normalized counts heatmap plotted along the
phylogeny, illustrating the phylogenetic structure of the differentially
expressed OGs.</p>
<div class="figure" style="text-align: center">
<img src="phylocompcodeR_check_figure/maplot-phyloHeatmap-1.png" alt="Example figures from the summarization report generated for a simulated data set. The tips colored by true differential expression status. Only the first 400 genes are represented. The first block of 200 genes are differencially expressed between condition 1 and 2. The second block of 200 genes are not differencially expressed." width="324"><p class="caption">
Example figures from the summarization report generated for a simulated
data set. The tips colored by true differential expression status. Only
the first 400 genes are represented. The first block of 200 genes are
differencially expressed between condition 1 and 2. The second block of
200 genes are not differencially expressed.
</p>
</div>
</div>
<div class="section level3">
<h3 id="performing-differential-expression-analysis">Performing differential expression analysis<a class="anchor" aria-label="anchor" href="#performing-differential-expression-analysis"></a>
</h3>
<p>Differential expression analysis can be conducted using the same
framework used in the <a href="compcodeR.html#performing-differential-expression-analysis"><code>compcodeR</code>
package</a>, through the <code>runDiffExp</code> function.</p>
<p>All the standard methods can be used. To account for the phylogenetic
nature of the data and for the varying length of the OGs, some methods
have been added to the pool.</p>
<p>The code below applies three differential expression methods to the
data set generated above: the <em><a href="https://bioconductor.org/packages/3.18/DESeq2" class="external-link">DESeq2</a></em>
method adapted for varying lengths, the <code>log2(TPM)</code>
transformation for length normalization, combined with <em><a href="https://bioconductor.org/packages/3.18/limma" class="external-link">limma</a></em>,
using the <code>trend</code> empirical Bayes correction, and accounting
for species-related correlations, and the phylogenetic regression tool
<a href="https://CRAN.R-project.org/package=phylolm" class="external-link"><code>phylolm</code></a>
applied on the same <code>log2(TPM)</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"alt_BM_repl1.rds"</span>,</span>
<span>           result.extent <span class="op">=</span> <span class="st">"DESeq2"</span>, Rmdfunction <span class="op">=</span> <span class="st">"DESeq2.createRmd"</span>,</span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>,</span>
<span>           fit.type <span class="op">=</span> <span class="st">"parametric"</span>, test <span class="op">=</span> <span class="st">"Wald"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"alt_BM_repl1.rds"</span>,</span>
<span>           result.extent <span class="op">=</span> <span class="st">"lengthNorm.limma"</span>, Rmdfunction <span class="op">=</span> <span class="st">"lengthNorm.limma.createRmd"</span>,</span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>,</span>
<span>           norm.method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>           length.normalization <span class="op">=</span> <span class="st">"TPM"</span>,</span>
<span>           data.transformation <span class="op">=</span> <span class="st">"log2"</span>,</span>
<span>           trend <span class="op">=</span> <span class="cn">FALSE</span>, block.factor <span class="op">=</span> <span class="st">"id.species"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/runDiffExp.html">runDiffExp</a></span><span class="op">(</span>data.file <span class="op">=</span> <span class="st">"alt_BM_repl1.rds"</span>,</span>
<span>           result.extent <span class="op">=</span> <span class="st">"phylolm"</span>, Rmdfunction <span class="op">=</span> <span class="st">"phylolm.createRmd"</span>,</span>
<span>           output.directory <span class="op">=</span> <span class="st">"."</span>,</span>
<span>           norm.method <span class="op">=</span> <span class="st">"TMM"</span>,</span>
<span>           model <span class="op">=</span> <span class="st">"BM"</span>, measurement_error <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>           extra.design.covariates <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>           length.normalization <span class="op">=</span> <span class="st">"TPM"</span>,</span>
<span>           data.transformation <span class="op">=</span> <span class="st">"log2"</span><span class="op">)</span></span></code></pre></div>
<p>As for a regular <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
analysis, example calls are provided in the reference manual (see the
help pages for the <code>runDiffExp</code> function), and a list of all
available methods can be obtained with the <code><a href="../reference/listcreateRmd.html">listcreateRmd()</a></code>
function.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/listcreateRmd.html">listcreateRmd</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "baySeq.createRmd"              </span></span>
<span><span class="co">#&gt;  [2] "DESeq2.createRmd"              </span></span>
<span><span class="co">#&gt;  [3] "DESeq2.length.createRmd"       </span></span>
<span><span class="co">#&gt;  [4] "DSS.createRmd"                 </span></span>
<span><span class="co">#&gt;  [5] "EBSeq.createRmd"               </span></span>
<span><span class="co">#&gt;  [6] "edgeR.exact.createRmd"         </span></span>
<span><span class="co">#&gt;  [7] "edgeR.GLM.createRmd"           </span></span>
<span><span class="co">#&gt;  [8] "lengthNorm.limma.createRmd"    </span></span>
<span><span class="co">#&gt;  [9] "lengthNorm.sva.limma.createRmd"</span></span>
<span><span class="co">#&gt; [10] "logcpm.limma.createRmd"        </span></span>
<span><span class="co">#&gt; [11] "NBPSeq.createRmd"              </span></span>
<span><span class="co">#&gt; [12] "NOISeq.prenorm.createRmd"      </span></span>
<span><span class="co">#&gt; [13] "phylolm.createRmd"             </span></span>
<span><span class="co">#&gt; [14] "sqrtcpm.limma.createRmd"       </span></span>
<span><span class="co">#&gt; [15] "TCC.createRmd"                 </span></span>
<span><span class="co">#&gt; [16] "ttest.createRmd"               </span></span>
<span><span class="co">#&gt; [17] "voom.limma.createRmd"          </span></span>
<span><span class="co">#&gt; [18] "voom.ttest.createRmd"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="comparing-results-from-several-differential-expression-methods">Comparing results from several differential expression methods<a class="anchor" aria-label="anchor" href="#comparing-results-from-several-differential-expression-methods"></a>
</h3>
<p>Given that the <code>phyloCompData</code> object has the same
structure with respect to the slots added by the differential expression
analysis (see <a href="compcodeR.html#the-result-object">the result
object</a>, the procedure to compare results from several differential
expression methods is exactly the same as for a <code>compData</code>
object, and can be found in the <a href="compcodeR.html#comparing-results-from-several-differential-expression-methods">corresponding
section</a> section of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
vignette.</p>
</div>
</div>
<div class="section level2">
<h2 id="using-your-own-data">Using your own data<a class="anchor" aria-label="anchor" href="#using-your-own-data"></a>
</h2>
<p><a href="compcodeR.html#using-your-own-data">As for a
<code>compData</code> object</a>, it is still possible to input
user-defined data to produce a <code>phyloCompData</code> object for
differential expression methods comparisons. One only needs to provide
the additional information needed, that is the phylogenetic tree, and
the length matrix. The constructor method will make sure that the tree
is consistent with the count and length matrices, with the same
dimensions and consistent species names.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Phylogentic tree with replicates</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span>text <span class="op">=</span> <span class="st">"(((A1:0,A2:0,A3:0):1,B1:1):1,((C1:0,C2:0):1.5,(D1:0,D2:0):1.5):0.5);"</span><span class="op">)</span></span>
<span><span class="co">## Sample annotations</span></span>
<span><span class="va">sample.annotations</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  condition <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>,                 <span class="co"># Condition of each sample</span></span>
<span>  id.species <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"D"</span><span class="op">)</span> <span class="co"># Species of each sample</span></span>
<span>  <span class="op">)</span></span>
<span><span class="co">## Count Matrix</span></span>
<span><span class="va">count.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">8000</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Length Matrix</span></span>
<span><span class="va">length.matrix</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">matrix</a></span><span class="op">(</span><span class="fl">1000</span><span class="op">*</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fl">8000</span><span class="op">)</span>, <span class="fl">1000</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">## Names must match</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">count.matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">length.matrix</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sample.annotations</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="co">## Extra infos</span></span>
<span><span class="va">info.parameters</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"mydata"</span>, uID <span class="op">=</span> <span class="st">"123456"</span><span class="op">)</span></span>
<span><span class="co">## Creation of the object</span></span>
<span><span class="va">cpd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/phyloCompData.html">phyloCompData</a></span><span class="op">(</span>count.matrix <span class="op">=</span> <span class="va">count.matrix</span>,</span>
<span>                     sample.annotations <span class="op">=</span> <span class="va">sample.annotations</span>,</span>
<span>                     info.parameters <span class="op">=</span> <span class="va">info.parameters</span>,</span>
<span>                     tree <span class="op">=</span> <span class="va">tree</span>,</span>
<span>                     length.matrix <span class="op">=</span> <span class="va">length.matrix</span><span class="op">)</span></span>
<span><span class="co">## Check</span></span>
<span><span class="fu"><a href="../reference/check_phyloCompData.html">check_phyloCompData</a></span><span class="op">(</span><span class="va">cpd</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="providing-your-own-differential-expression-code">Providing your own differential expression code<a class="anchor" aria-label="anchor" href="#providing-your-own-differential-expression-code"></a>
</h2>
<p>To use your own differential expression code, you can follow the <a href="compcodeR.html#providing-your-own-differential-expression-code">base
<code>compcodeR</code> instructions</a> in the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
vignette.</p>
</div>
<div class="section level2">
<h2 id="the-extended-data-object">The extended data object<a class="anchor" aria-label="anchor" href="#the-extended-data-object"></a>
</h2>
<p>The <code>phylocompData</code> data object is an S4 object that
extends <a href="compcodeR.html#the-data-object">the
<code>compData</code> object</a>, with the following added slots:</p>
<ul>
<li><p><code>tree</code> [<code>class phylo</code>]
(<strong>mandatory</strong>) – the phylogenetic tree describing the
relationships between samples.</p></li>
<li><p><code>length.matrix</code> [<code>class matrix</code>]
(<strong>mandatory</strong>) – the OG length matrix, with rows
representing genes and columns representing samples.</p></li>
<li>
<p>When produced with <code>generateSyntheticData</code>, the
<code>sample.annotations</code> data frame has added column:</p>
<ul>
<li>
<code>id.species</code> [<code>class character</code> or
<code>numeric</code>] – the species for each sample. Should match with
the <code>tip.label</code> of the <code>tree</code> slot.</li>
</ul>
</li>
<li>
<p>When produced with <code>generateSyntheticData</code>, the
<code>variable.annotations</code> data frame has an added columns:</p>
<ul>
<li>
<code>lengths.relmeans</code> [<code>class numeric</code>] – the
true mean values used in the simulations of the OG lengths.</li>
<li>
<code>lengths.dispersions</code> [<code>class numeric</code>] – the
true dispersion values used in the simulations of the OG lengths.</li>
<li>
<code>M.value.TPM</code> [<code>class numeric</code>] – the
estimated log2-fold change between conditions 1 and 2 for each OG using
TPM length normalization.</li>
<li>
<code>A.value.TPM</code> [<code>class numeric</code>] – the
estimated average expression in conditions 1 and 2 for each OG using TPM
length normalization.</li>
<li>
<code>prop.var.tree</code> [<code>class numeric</code>] – the
proportion of the variance explained by the phylogeny for each
gene.</li>
</ul>
</li>
</ul>
<p>The same way as the <code>compData</code> object, the
<code>phyloCompData</code> object needs to be saved to a file with
extension <code>.rds</code>.</p>
</div>
<div class="section level2">
<h2 id="the-evaluation-metrics">The evaluation metrics<a class="anchor" aria-label="anchor" href="#the-evaluation-metrics"></a>
</h2>
<p>The evaluation metrics are unchanged, and described in the <a href="compcodeR.html#the-evaluation-metrics">corresponding section</a>
section of the <em><a href="https://bioconductor.org/packages/3.18/compcodeR" class="external-link">compcodeR</a></em>
vignette.</p>
</div>
<div class="section level2">
<h2 id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt; Platform: x86_64-apple-darwin20 (64-bit)</span></span>
<span><span class="co">#&gt; Running under: macOS Monterey 12.6.6</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">#&gt; LAPACK: /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt; [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: internal</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">#&gt; [6] methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ape_5.7-1        compcodeR_1.37.0 sm_2.2-5.7.1    </span></span>
<span><span class="co">#&gt; [4] BiocStyle_2.29.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;  [1] gtable_0.3.3        xfun_0.39          </span></span>
<span><span class="co">#&gt;  [3] bslib_0.5.0         ggplot2_3.4.2      </span></span>
<span><span class="co">#&gt;  [5] caTools_1.18.2      lattice_0.21-8     </span></span>
<span><span class="co">#&gt;  [7] vctrs_0.6.3         tools_4.3.1        </span></span>
<span><span class="co">#&gt;  [9] bitops_1.0-7        generics_0.1.3     </span></span>
<span><span class="co">#&gt; [11] parallel_4.3.1      tibble_3.2.1       </span></span>
<span><span class="co">#&gt; [13] fansi_1.0.4         highr_0.10         </span></span>
<span><span class="co">#&gt; [15] cluster_2.1.4       pkgconfig_2.0.3    </span></span>
<span><span class="co">#&gt; [17] KernSmooth_2.23-21  desc_1.4.2         </span></span>
<span><span class="co">#&gt; [19] lifecycle_1.0.3     compiler_4.3.1     </span></span>
<span><span class="co">#&gt; [21] stringr_1.5.0       textshaping_0.3.6  </span></span>
<span><span class="co">#&gt; [23] gplots_3.1.3        munsell_0.5.0      </span></span>
<span><span class="co">#&gt; [25] clue_0.3-64         htmltools_0.5.5    </span></span>
<span><span class="co">#&gt; [27] sass_0.4.6          rmutil_1.1.10      </span></span>
<span><span class="co">#&gt; [29] yaml_2.3.7          pillar_1.9.0       </span></span>
<span><span class="co">#&gt; [31] pkgdown_2.0.7.9000  jquerylib_0.1.4    </span></span>
<span><span class="co">#&gt; [33] MASS_7.3-60         cachem_1.0.8       </span></span>
<span><span class="co">#&gt; [35] limma_3.57.6        vioplot_0.4.0      </span></span>
<span><span class="co">#&gt; [37] rpart_4.1.19        nlme_3.1-162       </span></span>
<span><span class="co">#&gt; [39] fBasics_4022.94     gtools_3.9.4       </span></span>
<span><span class="co">#&gt; [41] tidyselect_1.2.0    locfit_1.5-9.8     </span></span>
<span><span class="co">#&gt; [43] digest_0.6.31       stringi_1.7.12     </span></span>
<span><span class="co">#&gt; [45] dplyr_1.1.2         purrr_1.0.1        </span></span>
<span><span class="co">#&gt; [47] bookdown_0.34       rprojroot_2.0.3    </span></span>
<span><span class="co">#&gt; [49] fastmap_1.1.1       grid_4.3.1         </span></span>
<span><span class="co">#&gt; [51] colorspace_2.1-0    cli_3.6.1          </span></span>
<span><span class="co">#&gt; [53] magrittr_2.0.3      utf8_1.2.3         </span></span>
<span><span class="co">#&gt; [55] stable_1.1.6        edgeR_3.43.7       </span></span>
<span><span class="co">#&gt; [57] ROCR_1.0-11         scales_1.2.1       </span></span>
<span><span class="co">#&gt; [59] rmarkdown_2.22      timeDate_4022.108  </span></span>
<span><span class="co">#&gt; [61] png_0.1-8           zoo_1.8-12         </span></span>
<span><span class="co">#&gt; [63] timeSeries_4030.106 ragg_1.2.5         </span></span>
<span><span class="co">#&gt; [65] statip_0.2.3        memoise_2.0.1      </span></span>
<span><span class="co">#&gt; [67] evaluate_0.21       knitr_1.43         </span></span>
<span><span class="co">#&gt; [69] tcltk_4.3.1         stabledist_0.7-1   </span></span>
<span><span class="co">#&gt; [71] modeest_2.4.0       markdown_1.7       </span></span>
<span><span class="co">#&gt; [73] rlang_1.1.1         spatial_7.3-16     </span></span>
<span><span class="co">#&gt; [75] Rcpp_1.0.10         glue_1.6.2         </span></span>
<span><span class="co">#&gt; [77] BiocManager_1.30.21 jsonlite_1.8.5     </span></span>
<span><span class="co">#&gt; [79] R6_2.5.1            systemfonts_1.0.4  </span></span>
<span><span class="co">#&gt; [81] fs_1.6.2</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-Bastide2022" class="csl-entry">
Bastide, Paul, Charlotte Soneson, Olivier Lespinet, and Mélina Gallopin.
2022. <span>“Benchmark of Differential Gene Expression Analysis Methods
for Inter-Species RNA-Seq Data Using a Phylogenetic Simulation
Framework.”</span> <em>bioRxiv Preprint</em>. <a href="https://doi.org/10.1101/2022.01.21.476612" class="external-link">https://doi.org/10.1101/2022.01.21.476612</a>.
</div>
<div id="ref-Soneson2013" class="csl-entry">
Soneson, Charlotte, and Mauro Delorenzi. 2013. <span>“A Comparison of
Methods for Differential Expression Analysis of <span class="nocase">RNA-seq</span> Data.”</span> <em>BMC Bioinformatics</em>
14: 91.
</div>
<div id="ref-Stern2017" class="csl-entry">
Stern, David B., Jesse Breinholt, Carlos Pedraza-Lara, Marilú
López-Mejía, Christopher L. Owen, Heather Bracken-Grissom, James W.
Fetzner, and Keith A. Crandall. 2017. <span>“Phylogenetic Evidence from
Freshwater Crayfishes That Cave Adaptation Is Not an Evolutionary
Dead-End.”</span> <em>Evolution</em> 71 (10): 2522–32. <a href="https://doi.org/10.1111/evo.13326" class="external-link">https://doi.org/10.1111/evo.13326</a>.
</div>
<div id="ref-Stern2018" class="csl-entry">
Stern, David B., and Keith A. Crandall. 2018. <span>“The Evolution of
Gene Expression Underlying Vision Loss in Cave Animals.”</span>
<em>Molecular Biology and Evolution</em> 35 (8): 2005–14. <a href="https://doi.org/10.1093/molbev/msy106" class="external-link">https://doi.org/10.1093/molbev/msy106</a>.
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Charlotte Soneson, Paul Bastide, Mélina Gallopin.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
